### PR TITLE
Find an appropriate key based on the given KID and JWA

### LINKED ISSUE #20847: Private key JWT authentication no longer works on Keycloak 21
### Before reporting an issue

- [X] I have searched existing issues
- [X] I have reproduced the issue with the latest release

### Area

authentication

### Describe the bug

After upgrading from Keycloak 20.0.5 to Keycloak 21.1.1, private-key JWT authentication fails for clients whose credentials were set using a certificate.

### Version

21.1.1

### Expected behavior

Private key authentication works as it did in prior versions, given that no changes were announced in the release notes.

### Actual behavior

Attempting to authenticate returns an error `invalid_client (Unable to load public key)` and Keycloak logs the following message: `2023-06-07 07:58:32,312 WARN  [org.keycloak.keys.infinispan.InfinispanPublicKeyStorageProvider] (executor-thread-0) PublicKey wasn't found in the storage. Requested kid: 'null' . Available kids: '[YqQVzk2djJhvKyQtiEDl9FQ6iro8HiFXwl1ZdAiONeM]'`

### How to Reproduce?

1. Start a Keycloak 21 instance, e.g. using Docker.
2. Provision a new private key and certificate, e.g. using `openssl req -x509 -newkey ec -pkeyopt ec_paramgen_curve:P-521 -keyout client.key -nodes -sha512 -days 365 -out client.crt -subj '/CN=test-client'`.
3. Create a new OIDC client, e.g. "foo", in Keycloak with service account enabled.
4. Set its credential type to "Signed JWT" and import the certificate created earlier.
5. Attempt to authenticate as the client using the following NodeJS script (requires `jose` and `openid-client` NPM packages):
```js
import {readFile} from "node:fs/promises";
import {exportJWK, importPKCS8} from "jose";
import {Issuer} from "openid-client";

const privateKey = await importPKCS8(await readFile("./client.key", "utf8"), "ES512");
const issuer = await Issuer.discover("http://localhost:8080/realms/master");
const client = new issuer.Client(
  {
    client_id: "foo",
    token_endpoint_auth_method: "private_key_jwt"
  },
  { keys: [await exportJWK(privateKey)] }
);
const {access_token: accessToken} = await client.grant({ grant_type: "client_credentials" });
console.log(accessToken);
```

The same works on 20.0.5.

### Anything else?

It appears that Keycloak internally computes a `kid` and requires it from then on. After looking through the diff between 20.0.5 and 21.1.1, I believe #15114 might be responsible.

### PR DESCRIPTION
**Pull request by patrickjennings**
#20847

Changes the logic when searching for an appropriate key to use given a KID and JWA.

Prefers matching both KID and alg. Then matching against kid. Followed by alg. If none is found, returns the first key in the list.

### CODE REVIEW

**Review by pedroigor — COMMENTED** [review]
@patrickjennings Does it make more sense to make very explicit the key that should be used as a fallback in case no `kid` is set?

For instance, mark the key here https://github.com/keycloak/keycloak/blob/2ba5ca3c5f3a4163a8ef4c2bad16445cdce10ee3/services/src/main/java/org/keycloak/keys/loader/ClientPublicKeyLoader.java#L116 that it was built from a client certificate.

And then use this marker to resolve the default key.

**Comment by mposolda** [comment]
> @patrickjennings Does it make more sense to make very explicit the key that should be used as a fallback in case no `kid` is set?
> 
> For instance, mark the key here
> 
> https://github.com/keycloak/keycloak/blob/2ba5ca3c5f3a4163a8ef4c2bad16445cdce10ee3/services/src/main/java/org/keycloak/keys/loader/ClientPublicKeyLoader.java#L116
> that it was built from a client certificate.
> 
> And then use this marker to resolve the default key.

+1 to this proposal (if possible to implement without too much complexity)

**Comment by patrickjennings** [comment]
@pedroigor , @mposolda , I added the explicit fallback on the assigned cert for the client if built from the client cert. That worked really well.

I also extended the logic for the keys cache in the InfinispanPublicKeyStorageProvider. Basically, if KID is not provided (null), we can partially match on alg only if the cache has been refreshed recently. Otherwise, we want to refresh the cache and then we can partially match on the alg.

These changes fix the problem in keycloak-20847. Let me know what you think of these additions.

File: js/apps/admin-ui/public/locales/en/clients.json [hunk 1]
```diff
@@ -519,7 +519,7 @@
   "browserFlow": "Browser Flow",
   "directGrant": "Direct Grant Flow",
   "jwksUrlConfig": "JWKS URL configs",
-  "keysIntro": "If \"Use JWKS URL switch\" is on, you need to fill a valid JWKS URL. After saving, admin can download keys from the JWKS URL or keys will be downloaded automatically by Keycloak server when see the stuff signed by the unknown KID",
+  "keysIntro": "If \"Use JWKS URL switch\" is on, you need to fill a valid JWKS URL. After saving, admin can download keys from the JWKS URL or keys will be downloaded automatically by Keycloak server when an unknown KID is seen during authentication.",
```
Comment (mposolda) [inline js/apps/admin-ui/public/locales/en/clients.json:522]:
```suggestion
  "keysIntro": "If \"Use JWKS URL switch\" is on, you need to fill a valid JWKS URL. After saving, admin can download keys from the JWKS URL or keys will be downloaded automatically by Keycloak server when an unknown KID is seen during client authentication.",
```

File: model/infinispan/src/main/java/org/keycloak/keys/infinispan/InfinispanPublicKeyStorageProvider.java [hunk 2]
```diff
@@ -124,22 +124,23 @@ public KeyWrapper getFirstPublicKey(String modelKey, String algorithm, PublicKey
 
     @Override
     public KeyWrapper getPublicKey(String modelKey, String kid, String algorithm, PublicKeyLoader loader) {
-        // Check if key is in cache
         PublicKeysEntry entry = keys.get(modelKey);
-        if (entry != null) {
+        int lastRequestTime = entry==null ? 0 : entry.getLastRequestTime();
+        int currentTime = Time.currentTime();
+        boolean isKeyCacheStale = currentTime > lastRequestTime + minTimeBetweenRequests;
```
Comment (mposolda) [inline model/infinispan/src/main/java/org/keycloak/keys/infinispan/InfinispanPublicKeyStorageProvider.java:130]: Minor: Maybe this boolean variable can be renamed to `isSendingRequestAllowed` ? As the cache entry is technically not state?

**Review by mposolda — DISMISSED** [review]
@patrickjennings Thanks for the fix! I've added 2 minor comments inline. Besides that, some additional concerns:
- Can you add automated test for this scenario? Maybe adding something to class `ClientAuthSignedJWTTest` is sufficient. For instance see this old PR for the inspiration https://github.com/keycloak/keycloak/pull/15114/files

- Did you check also class `MapPublicKeyStorageProvider`? It's possible there are not changes needed in this class, but I am not 100% sure (It is part of the new storage when the `InfinispanPublicKeyStorageProvider` is part of the old storage). Hopefully when you add new test and everything works, we're likely OK (Leaving to you to doublecheck...)

**Review by pedroigor — DISMISSED** [review]

**Comment by pedroigor** [comment]
@patrickjennings LGTM. There is a comment from @mposolda about a variable name. Not sure if he is hard on it.

**Review by pedroigor — DISMISSED** [review]

**Comment by patrickjennings** [comment]
@mposolda Thanks for the review and suggestions. Made the necessary changes. And added an additional integration test specifically with the parameters detailed in #21160 .

Not too sure about how to run with the new store. Are we able to run the integration tests with the new store to check correctness? @pedroigor , do you know?

**Comment by pedroigor** [comment]
@patrickjennings I don't think we need to consider running this one using the new store.

**Comment by jonkoops** [comment]
Since this PR only updates some labels in the UI, I am removing the UI team from this as reviewers. I think we can trust y'all :wink:

**Review by pedroigor — APPROVED** [review]

**Pull request merged**